app/src/main/java/com/infomaniak/mail/utils/extensions/Extensions.kt

```diff
@@ -334,20 +319,37 @@ fun List<Folder>.flattenFolderChildren(dismissHiddenChildren: Boolean = false):
             val children = with(folder.children) {
                 (if (dismissHiddenChildren) query("${Folder::isHidden.name} == false") else query())
                     .sort(Folder::name.name, Sort.ASCENDING)
+                    .sort(Folder::isFavorite.name, Sort.DESCENDING)
                     .find()
             }
-            inputList.addAll(0, children)
+            inputList.addAll(index = 0, children)
         } else {
             outputList.add(folder)
-            val children = with(folder.children) { if (dismissHiddenChildren) filter { !it.isHidden } else this }
-            inputList.addAll(children)
+            val children = (if (dismissHiddenChildren) folder.children.filter { !it.isHidden } else folder.children)
+                .sortedBy { it.name }
+                .sortedBy { !it.isFavorite }
+            inputList.addAll(index = 0, children)
         }
 
         return if (inputList.isEmpty()) outputList else formatFolderWithAllChildren(inputList, outputList)
     }
 
     return formatFolderWithAllChildren(toMutableList())
 }
+
+fun List<Folder>.addDividerBeforeFirstCustomFolder(dividerType: Any): List<Any> {
+    val folders = this
+    val items = mutableListOf<Any>()
+    var needsToAddDivider = true
+    folders.forEach { folder ->
+        if (needsToAddDivider && folder.isRootAndCustom) {
+            needsToAddDivider = false
+            items.add(dividerType)
+        }
+        items.add(folder)
+    }
+    return items
+}
```

Comment:
You could technically use buildList to create and return this part instead of casting a mutable list to a list:

```kt
fun List<Folder>.addDividerBeforeFirstCustomFolder(dividerType: Any): List<Any> {
    val folders = this
    var needsToAddDivider = true
    return buildList {
        folders.forEach { folder ->
            if (needsToAddDivider && folder.isRootAndCustom) {
                needsToAddDivider = false
                add(dividerType)
            }
            add(folder)
        }
    }
}
```